### PR TITLE
Fix thread handle leak on Windows

### PR DESCRIPTION
**Pull request by kankri**
* Tentative fix for issue #398

* Removed unused 'resolver_thread' in 'struct host_entry'
   * `host_entry->resolver_thread` was used only in function `create_and_init_host_entry()`, converted it to a local variable.
* Clean up 'resolver_thread' to avoid leaking a thread handle on Windows
   * The thread handle inside `struct aws_thread` was never closed, calling `aws_thread_clean_up()` fixes this.
   * There is a copy of the thread struct in `thread_wrapper` with a unique thread handle which is used to synchronize thread exit and it gets properly closed.
   * I noticed that the semantics of `aws_thread_clean_up()` on Linux are a bit different: it calls `pthread_detach()`. Not sure if this matters.

----
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/host_resolver.c [hunk 1]
```diff
@@ -1360,15 +1361,17 @@ static inline int create_and_init_host_entry(
     struct aws_thread_options thread_options = *aws_default_thread_options();
     thread_options.join_strategy = AWS_TJS_MANAGED;
 
-    aws_thread_launch(&new_host_entry->resolver_thread, resolver_thread_fn, new_host_entry, &thread_options);
+    aws_thread_launch(&resolver_thread, resolver_thread_fn, new_host_entry, &thread_options);
```
Comment (TingDaoK) [inline source/host_resolver.c:1364]: Thank you for your PR.

> You must either call join or detach after creating the thread and before calling clean_up.

Referring to [here](https://github.com/awslabs/aws-c-common/blob/main/include/aws/common/thread.h#L119). I think we cannot just make the thread a local variable and clean it up like that. But seems like we do forget about thread clean up when clean up the host_entry. Not super sure, still investigating.

Comment (TingDaoK) [reply source/host_resolver.c:1364]: Maybe instead of making it a local variable, add `aws_thread_clean_up(&entry->resolver_thread)` to `s_clean_up_host_entry`?

**Comment by bretambrose** [comment]
I think we've been using aws_thread_clean_up in an inconsistent fashion (timing/lifecycle wise) within the codebase and while addressing the core issue (handle leak) that inconsistency should be resolved as well.

We want to factor in the managed thread concept (which was created to allow for threads that weren't "owned" by a particular object but were still guaranteed to be joined somehow, somewhere given a core API call (aws_thread_join_all_managed)).

I also worry about the caveat in https://docs.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject ("If this handle is closed while the wait is still pending, the function's behavior is undefined.")  It's not 100% clear if the undefined behavior is limited to a handle that was closed while the Wait was happening or also before the Wait was called.  But just to be cautious, I would assume that a Wait on a closed handle was potentially dangerous too (although it's true we are doing that in places right now).

So here's what I'm currently toying with as a contract:

For managed threads (host resolver threads and event loop group async shutdown threads):
   The managed thread logic will invoke aws_thread_clean_up after the thread has been joined (https://github.com/awslabs/aws-c-common/blob/main/source/windows/thread.c#L56).  This would remove a call to aws_thread_clean_up in https://github.com/awslabs/aws-c-io/blob/main/source/event_loop.c#L76 and cause the host resolver threads to get clean up called on them.  I've done some preliminary testing and it looks positive so far.

For unmanaged threads (everything else):
   aws_thread_clean_up must be called and must only be called after any desired join behavior

The consequences I see is that we're doing things backwards in a number of our tests which will now need to call cleanup at the end of their work rather than in the middle.

File: source/host_resolver.c (hunk 1 above)
Comment (kankri) [reply source/host_resolver.c:1364]: It seemed to me that when creating a managed thread calling join and clean up will be taken care of by the thread wrapper...

**Comment by kankri** [comment]
>  I had forgotten that we actually allocate multiple Windows handles when launching a thread: one is the handle returned by CreateThread, but the other is a handle opened within the thread's wrapper function so that it can be joined eventually by the managed thread logic once the thread function completes.

Yes, this is what I was referring to by this in my description:
> There is a copy of the thread struct in thread_wrapper with a unique thread handle which is used to synchronize thread exit and it gets properly closed.

Regarding:
> "If this handle is closed while the wait is still pending, the function's behavior is undefined."

Since the copy of the handle in the wrapper is used for safe thread joining as far as I can see, it shouldn't matter if the handle returned to the creator of the thread is closed: since it is never used for anything, it is not used for WaitForSingleThread() either.

I believe my change causes the correct amount of CloseHandle()s and thread handle waiting and joining, but I can't claim I fully understand all the details of managed vs non-managed threads in the SDK. And I was also worried if calling aws_thread_clean_up () in Linux has some different effects than on Windows (where it just closes a handle but doesn't otherwise affect the thread).

So I made this PR just in case the simple fix is OK, but if it isn't so, please go ahead and make an alternative PR.

**Comment by bretambrose** [comment]
> I believe my change causes the correct amount of CloseHandle()s and thread handle waiting and joining, but I can't claim I fully understand all the details of managed vs non-managed threads in the SDK. And I was also worried if calling aws_thread_clean_up () in Linux has some different effects than on Windows (where it just closes a handle but doesn't otherwise affect the thread).
> 
> So I made this PR just in case the simple fix is OK, but if it isn't so, please go ahead and make an alternative PR.

Your PR absolutely fixes the host resolver thread handle leak; there's nothing wrong with it.  But from a more general perspective, it's suddenly clear that how you (or anyone at the C layer) should use aws_thread_clean_up() is very unclear and I would like the full solution to make that consistent as well.

My intent with the managed threads is that they should be something that you just create and forget about, including both joins and clean up (detach).  I think it's unfortunate that the original thread api didn't distinguish between detach and clean up and instead conflated the two but that's what we have to live with now I guess.

So, in summary, issue is spot-on, analysis and verification steps spot-on, PR fixes the issue, but I would like to go with mine because I think it makes the thread contract clearer (otherwise everyone using a managed thread still needs to call clean up after launching even though clean up is considered an end-of-lifecycle operation everywhere in the aws-c-* libraries).  We'll see what others says in review; I certainly understand that an either/or contract (if managed do this else do that) is not optimal.

**Comment by kankri** [comment]
Yes, I agree that it makes sense to clarify the contract and it is redundant for `aws_thread_launch()` to return an "active" thread object for the caller when creating a managed thread. Your PR awslabs/aws-c-common#814 solves this by calling `aws_thread_clean_up()` before returning from `aws_thread_launch()`. So I'm happy if this is closed in favor of your 2 pull requests.

**Comment by bretambrose** [comment]
Thanks again for tracking this down.  I've merged the hopefully-clarifying multi-repo fix for this in common and io and made new releases.  The C++ CRT release may be delayed slightly (a day or two) in order to bring in other fixes.
